developer-docs/architecture/type-system.md

```diff
@@ -0,0 +1,201 @@
+# Pulumi Type System
+
+In its role as a broker of information between various actors--e.g. language SDKs,
+resource providers, multi-language components, and statefiles--and in its role as a
+programming model, it is important that Pulumi deals in values with well-defined
+semantics. The _Pulumi type system_ specifies these semantics. It is the responsibility of
+each language SDK and interchange format to ensure that these semantics are faithfully
+implemented, ideally in as idiomatic a fashion as possible.
+
+## Primitive Types
+
+The core primitives of the Pulumi type system form a superset of the JSON type system, and supports the
+following types:
+
+- `Null`, which represents the lack of a value
+- `Bool`, which represents a boolean value
+- `Number`, which represents an IEEE-754 double-precision number
+- `String`, which represents a sequence of UTF-8 encoded unicode code points
+- [`Asset`](#assets-and-archives), which represents a blob
+- [`Archive`](#assets-and-archives), which represents a map from strings to `Asset`s or
+  `Archive`s
+- [`ResourceReference`](#resource-references), which represents a reference to a resource
+- `Array<T>`, which represents a numbered sequence of values of a particular type
+- `Map<T>`, which represents an unordered mapping from strings to values of a particular
+  type
+- `Union<T0, T1 ... TN>`, which represents a value of one of a fixed set of types
+- `Enum<T, V0 ... VN>`, which represents one of a fixed set of values of a particular type
+
+### Assets and Archives
+
+An `Asset` or `Archive` may contain either literal data or a reference to a file or URL.
+In the former case, the literal data is a textual string or a map from strings to `Asset`s
+or `Archive`s, respectively. In the latter case, the referenced file or URL is an opaque
+blob or a TAR, gzipped TAR, or ZIP archive, respectively.
```

Comment:
I find this paragraph a little difficult to follow because of the distinction of `Asset` vs `Archive` mixed with the distinction between `literal data` and `reference to a fail or URL`.

Maybe something like the following is easier to follow?
```suggestion
An `Asset` or `Archive` may contain either literal data or a reference to a file or URL.

* Literal data:
	* Asset: a text string
	* Archive: a map from strings to `Asset`s or other `Archive`s
* Reference to a file or URL:
	* Asset: Opaque blob
	* Archive: TAR, gzipped TAR or ZIP archive
```

Reply:
Maybe even a table 🤷🏽 ?

Reply:
Rewrote it as two paragraphs: one for assets and one for archives. PTAL

Reply:
👍🏽